iox_query/src/plan/influxql/rewriter.rs

```diff
@@ -0,0 +1,829 @@
+                        ));
+                    }
+
+                    new_fields.push(f.clone());
+                }
+
+                _ => new_fields.push(f.clone()),
+            }
+        }
+
+        stmt.fields = FieldList::new(new_fields);
+    }
+
+    if has_group_by_wildcard {
+        let group_by_tags = if has_group_by_wildcard {
+            tag_set.into_iter().sorted().collect::<Vec<_>>()
+        } else {
+            vec![]
+        };
+
+        if let Some(group_by) = &stmt.group_by {
+            let mut new_dimensions = Vec::new();
+
+            for dim in group_by.iter() {
+                let add_dim = |dim: &String| {
+                    new_dimensions.push(Dimension::Tag(Identifier::new(dim.clone())))
+                };
+
+                match dim {
+                    Dimension::Wildcard => {
+                        group_by_tags.iter().for_each(add_dim);
+                    }
+                    Dimension::Regex(re) => {
+                        let re = parse_regex(re)?;
+
+                        group_by_tags
+                            .iter()
+                            .filter(|dim| re.is_match(dim.as_str()))
+                            .for_each(add_dim);
+                    }
+                    _ => new_dimensions.push(dim.clone()),
+                }
+            }
+            stmt.group_by = Some(GroupByClause::new(new_dimensions));
+        }
+    }
+
+    Ok(())
+}
+
+/// Recursively rewrite the specified [`SelectStatement`], expanding any wildcards or regular expressions
+/// found in the projection list, `FROM` clause or `GROUP BY` clause.
+pub(crate) fn rewrite_statement(
+    q: &SelectStatement,
+    schema: Arc<dyn SchemaProvider>,
+) -> Result<SelectStatement> {
+    let mut stmt = q.clone();
+    rewrite_from(&mut stmt, Arc::clone(&schema))?;
+    rewrite_field_list(&mut stmt, schema)?;
+
+    Ok(stmt)
+}
+
+#[cfg(test)]
+mod test {
+    use crate::plan::influxql::rewriter::{has_wildcards, rewrite_statement, walk_expr_mut};
+    use crate::plan::influxql::test_utils::{get_first_field, MockSchemaProvider};
+    use influxdb_influxql_parser::expression::Expr;
+    use influxdb_influxql_parser::literal::Literal;
+    use influxdb_influxql_parser::parse_statements;
+    use influxdb_influxql_parser::select::SelectStatement;
+    use influxdb_influxql_parser::statement::Statement;
+    use test_helpers::assert_contains;
+
+    fn parse_select(s: &str) -> SelectStatement {
+        let statements = parse_statements(s).unwrap();
+        match statements.first() {
+            Some(Statement::Select(sel)) => *sel.clone(),
+            _ => unreachable!(),
+        }
+    }
+
+    #[test]
+    fn test_rewrite_statement() {
+        // Exact, match
+        let stmt = parse_select("SELECT usage_user FROM cpu");
+        let stmt = rewrite_statement(&stmt, MockSchemaProvider::new_schema_provider()).unwrap();
+        assert_eq!(format!("{}", stmt), "SELECT usage_user::float FROM cpu");
```

Comment:
A style suggestion is that you can get the string form using `to_string()` rather than `format!()` like

```suggestion
        assert_eq!(stmt.to_string(), "SELECT usage_user::float FROM cpu");
```

I don't think it matters but I wanted to point it out

Reply:
Definitely cleaner and thanks for pointing it out. I see that the implementation of `ToString` is generic for any type that implements `fmt::Display`, which is nice.